app/src/main/java/org/worldbank/transport/driver/services/DriverLocationService.java

```diff
@@ -0,0 +1,465 @@
+            return DriverLocationService.this;
+        }
+    }
+
+    private LocationManager locationManager;
+    private LocationListener locationListener;
+    private Context context;
+    private final IBinder binder = new LocationServiceBinder();
+    private DriverLocationUpdateListener driverLocationUpdateListener;
+
+    // will be non-null when at least one location update received with the minimum accuracy
+    private Location acceptableLocation;
+    private long startedWaitingForBetterLocation;
+    private ArrayList<Location> preferredAccuracyLocations;
+    private boolean doneWaitingForLocations;
+
+    @Override
+    public void onConnected(@Nullable Bundle bundle) {
+        // Android Studio will complain mightily if requestLocationUpdates is not called
+        // after a permissions check that is in the same method
+        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
+            Log.d(LOG_LABEL, "Actually starting location updates now.");
+            locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, locationListener);
+        } else {
+            Log.w(LOG_LABEL, "Got onConnected callback, but do not have location services permissions now");
+        }
+    }
+
+    @Override
+    public void onConnectionSuspended(int i) {
+        Log.w(LOG_LABEL, "Location services API connection suspended!");
+        // TODO: anything?
+    }
+
+    @Override
+    public void onConnectionFailed(ConnectionResult connectionResult) {
+        Log.w(LOG_LABEL, "Location services API connection failed!");
+        // TODO: anything?
+    }
+
+    /**
+     * Return best location estimate found for received updates.
+     * Should stop this service after receiving this result (can be accomplished by unbinding)
+     *
+     * @return best estimate of current location during service run
+     */
+    @Nullable
+    public Location getBestLocationFound() {
+
+        if (preferredAccuracyLocations.size() > 0) {
+            // calculate weighted average location from those found
+
+            if (preferredAccuracyLocations.size() == 1) {
+                // if only got one, just use it
+                return new Location(preferredAccuracyLocations.get(0));
+            }
+
+            Log.d(LOG_LABEL, "Finding weighted average of " + preferredAccuracyLocations.size() + " location readings");
+
+            double sumLat = 0;
+            double sumLon = 0;
+            float sumAccuracy = 0;
+            double sumInvertedAccuracy = 0;
+
+            for (Location location : preferredAccuracyLocations) {
+                float accuracy = location.getAccuracy();
+
+                // accuracy for these readings is within range of 0 to PREFERRED_LOCATION_ACCURACY
+                // smaller is better; invert range for weighted average
+                float invertedAccuracy = PREFERRED_LOCATION_ACCURACY - accuracy;
```

Comment:
What does the accuracy represent? If it represents a radius in meters then I wonder if we need to scale the numbers here since the uncertainty about the point's actual location will change with the square of the radius accuracy. But I don't know much about how the location API works so I might be totally off here.